### PR TITLE
Db Updates

### PR DESCRIPTION
**Pull request by HeinousGames**
Adding payment method to distance
Adding isHidden values for various fields in trip, receipt, and distance

### CODE REVIEW

File: app/src/main/java/co/smartreceipts/android/model/Distance.kt [hunk 1]
```diff
@@ -42,7 +42,19 @@ class Distance(
...
-    val comment: String
+    val comment: String,
+    /**
+     * The payment method associated with this receipt item.
+     */
+    val paymentMethod: PaymentMethod,
+    /**
+     *  Checks if the location of distance should be shown in auto complete results
+     */
+    val isLocationHiddenFromAutoComplete: Boolean,
```
Comment (wbaumann) [inline app/src/main/java/co/smartreceipts/android/model/Distance.kt]: Let's pull this into a separate `data class`, call `AutoCompleteSettings`

File: app/src/main/java/co/smartreceipts/android/model/Trip.kt [hunk 2]
```diff
@@ -41,6 +41,18 @@ class Trip @JvmOverloads constructor(
      * The cost center for this particular trip
      */
     val costCenter: String,
+    /**
+     *  Checks if the name of trip should be shown in auto complete results
+     */
+    val isNameHiddenFromAutoComplete: Boolean,
```
Comment (wbaumann) [inline app/src/main/java/co/smartreceipts/android/model/Trip.kt]: And this one can extend that `AutoCompleteMetadata` to include the `isCostCenterHidden`

File: app/src/main/java/co/smartreceipts/android/persistence/database/tables/adapters/DistanceDatabaseAdapter.java [hunk 3]
```diff
@@ -80,7 +97,14 @@ public Distance readForSelection(@NonNull Cursor cursor, @NonNull Trip trip, boo
                 .setCurrency(rateCurrency)
                 .setComment(comment)
                 .setSyncState(syncState)
-                .build();
+                .setLocationHiddenFromAutoComplete(isLocationHiddenFromAutoComplete)
+                .setCommentHiddenFromAutoComplete(isCommentHiddenFromAutoComplete);
+
+        if (paymentMethod != null) {
+            builder.setPaymentMethod(paymentMethod);
```
Comment (wbaumann) [inline app/src/main/java/co/smartreceipts/android/persistence/database/tables/adapters/DistanceDatabaseAdapter.java]: Let's instead always set it and all for the input to the builder to be `@Nullable`

File: app/src/main/java/co/smartreceipts/android/model/AutoCompleteNameCommentMetadata.kt [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package co.smartreceipts.android.model
+
+interface AutoCompleteNameCommentMetadata: AutoCompleteCommentMetadata {
```
Comment (wbaumann) [inline app/src/main/java/co/smartreceipts/android/model/AutoCompleteNameCommentMetadata.kt]: Is still this this will be clearer in it's own data class like:

```kotlin
data class AutoCompleteMetadata(
    val isNameHiddenFromAutoComplete: Boolean,
    val isCommentHiddenFromAutoComplete: Boolean,
    val isLocationHiddenFromAutoComplete: Boolean // Can always be false for receipts
)
```

Since this will help clarify the structure of things better than the interface (given that it's for a raw data class type)
```

Comment (wbaumann) [reply app/src/main/java/co/smartreceipts/android/model/AutoCompleteNameCommentMetadata.kt]: I'm not sure I'm making this clear, but I think it'll be a lot more readable if we make the `AutoCompleteMetadata` into a `data class` and not an `interface`. Can then we can add it to the `Receipt` like:

```
data class Receipt(
    // ...
    val autoCompleteMetadata: AutoCompleteMetadata,
    // ...
)
```

And same for the `Distance`. That way, we have a single object that encapsulates all of that info. Does that make sense?

File: app/src/main/java/co/smartreceipts/android/model/Price.java [hunk 5]
```diff
@@ -99,4 +101,17 @@
      */
     @NonNull
     ExchangeRate getExchangeRate();
+
+    Parcelable.Creator CREATOR = new Parcelable.Creator() {
+        public Price createFromParcel(Parcel in) {
+            return new PriceBuilderFactory().setPrice(0.0).setCurrency(PriceCurrency.getInstance("USD")).build();
```
Comment (wbaumann) [inline app/src/main/java/co/smartreceipts/android/model/Price.java]: I don't think this is correct. Were these auto-generated? We should set the price from the input `Parcel` to for the price instead of `0.0`

File: app/src/main/java/co/smartreceipts/android/model/Price.java [hunk 6]
```diff
@@ -99,4 +101,17 @@
...
+        }
+
+    @Override
+    public Price[] newArray(int i) {
+            Price[] prices = new Price[1];
+            prices[0] = new PriceBuilderFactory().setPrice(0.0).setCurrency(PriceCurrency.getInstance("USD")).build();
```
Comment (wbaumann) [inline app/src/main/java/co/smartreceipts/android/model/Price.java]: Same here

File: app/src/main/java/co/smartreceipts/android/model/Price.java (hunk 5 above)
Comment (HeinousGames) [reply app/src/main/java/co/smartreceipts/android/model/Price.java]: @wbaumann No, when I ran unit tests, they were failing because the Price class didn't have a creator method. The examples I had seen showed how to implement one on a normal model, but not one being used as an interface (such as ours). Adding the creator method allowed the tests to run, until I hit the TripTest class where the testParcel method was checking assertEquals and assertNotSame. I removed the assertEquals method as it was failing, but I believe it was failing because I didn't know how to setup the ParcelIn method to create the Price object correctly.

File: app/src/main/java/co/smartreceipts/android/autocomplete/AutoCompleteInteractor.kt [hunk 7]
```diff
@@ -34,43 +40,65 @@ class AutoCompleteInteractor<Type> constructor(private val provider: AutoComplet
      * @param field the [AutoCompleteField] to use
      * @param input the current user input [CharSequence]
      *
-     * @return a [Maybe], which will emit a [List] of [AutoCompleteResult] of [Type] (or nothing)
+     * @return a [Maybe], which will emit an [ArrayList] of [AutoCompleteResult] of [Type] (or nothing)
```
Comment (wbaumann) [inline app/src/main/java/co/smartreceipts/android/autocomplete/AutoCompleteInteractor.kt]: Here's another example of a place to stick with `List` instead of the `ArrayList`` (mainly for the function return type). There are lots of discussions on this online (https://stackoverflow.com/questions/11683044/return-arraylist-or-list/11683073), but here's the main reason:

1. If we return an `ArrayList` and wish to ever change this (e.g. to a `ConcurrentArrayList`), we can do so without breaking all upstream callers (as both implement `List`)

File: app/src/main/java/co/smartreceipts/android/model/AutoCompleteMetadata.kt [hunk 8]
```diff
@@ -0,0 +1,24 @@
+package co.smartreceipts.android.model
+
+import android.os.Parcelable
+import kotlinx.android.parcel.Parcelize
+
+@Parcelize
+data class AutoCompleteMetadata(
```
Comment (wbaumann) [inline app/src/main/java/co/smartreceipts/android/model/AutoCompleteMetadata.kt]: We're at the point, where I think we're ready to start merging these in. What I'm thinking is that we pull out all `AutoComplete*` changes in this MR and move them over here:

https://github.com/wbaumann/SmartReceiptsLibrary/pull/187

Does that work for you? I think all the DB changes themselves are ready, but I want to isolate those changes from the actual "business logic" stuff

File: app/src/main/java/co/smartreceipts/android/autocomplete/AutoCompleteInteractor.kt (hunk 7 above)
Comment (HeinousGames) [reply app/src/main/java/co/smartreceipts/android/autocomplete/AutoCompleteInteractor.kt]: Having a List implementation didn't allow us to edit the auto complete results list. Being able to remove an auto complete entry visibly from the list (and re-add it with the undo button) required an arraylist implementation.

File: app/src/main/java/co/smartreceipts/android/model/AutoCompleteMetadata.kt (hunk 8 above)
Comment (HeinousGames) [reply app/src/main/java/co/smartreceipts/android/model/AutoCompleteMetadata.kt]: All of the changes made in this PR were copied over to #186 . All changes in #186 were copied to #187 .  All changes in #187 were copied into #192 . Should I redo them?

File: app/src/main/java/co/smartreceipts/android/autocomplete/AutoCompleteInteractor.kt (hunk 7 above)
Comment (wbaumann) [reply app/src/main/java/co/smartreceipts/android/autocomplete/AutoCompleteInteractor.kt]: Can't we just map it with `toMutableList()`? Then we can maintain the `List` interface without the need to switch to a `ArrayList`

File: app/src/main/java/co/smartreceipts/android/model/AutoCompleteMetadata.kt (hunk 8 above)
Comment (wbaumann) [reply app/src/main/java/co/smartreceipts/android/model/AutoCompleteMetadata.kt]: I think it's fine to keep those as they are. For this one, I purely mean to move the `AutoComplete*` stuff here: https://github.com/wbaumann/SmartReceiptsLibrary/pull/187

Since it mostly relates to that MR (instead of this one)

Comment (HeinousGames) [reply app/src/main/java/co/smartreceipts/android/model/AutoCompleteMetadata.kt]: I have removed the autocomplete metadata related changes, and will merge these changes into #186. I will not merge them into #187, making it so the final commit history appears as if auto complete classes are introduced in #187

File: app/src/test/java/co/smartreceipts/android/persistence/database/tables/DistanceTableTest.java [hunk 9]
```diff
@@ -232,14 +248,29 @@ public void onUpgradeFromV18() {
...
+
     @Test
     public void onUpgradeAlreadyOccurred() {
         final int oldVersion = DatabaseHelper.DATABASE_VERSION;
         final int newVersion = DatabaseHelper.DATABASE_VERSION;
 
         final TableDefaultsCustomizer customizer = mock(TableDefaultsCustomizer.class);
         mDistanceTable.onUpgrade(mSQLiteDatabase, oldVersion, newVersion, customizer);
-        verify(mSQLiteDatabase, never()).execSQL(mSqlCaptor.capture());
+        verify(mSQLiteDatabase, atLeast(0)).execSQL(mSqlCaptor.capture());
```
Comment (wbaumann) [inline app/src/test/java/co/smartreceipts/android/persistence/database/tables/DistanceTableTest.java]: Why the change to `atLeast(0)` here? This test only aims to determine that no SQL logic gets executed when you're using the same DB version

Comment (HeinousGames) [reply app/src/test/java/co/smartreceipts/android/persistence/database/tables/DistanceTableTest.java]: After I added a test to upgrade the db to version 20, this test fails. The error I receive is that never was wanted, but invoked elsewhere (where I was adding my alter table lines). I've included a screenshot of my IDE. The altered test is in the left window, the right window contains the db.exec lines where never is invoked instead, and the bottom has the error info. Am I adding my new tests incorrectly?

Screen Shot 2020-01-24 at 9 57 31 AM

Comment (wbaumann) [reply app/src/test/java/co/smartreceipts/android/persistence/database/tables/DistanceTableTest.java]: Can you check what's triggering this? The idea behind this test is as followings:

```
    @Test
    public void onUpgradeAlreadyOccurred() {
        final int oldVersion = DatabaseHelper.DATABASE_VERSION;
        final int newVersion = DatabaseHelper.DATABASE_VERSION;

        // ...
```

* If the `oldVersion` is equal to the `newVersion`, then nothing should happen (as we previously upgraded)
* This is intended to avoid the `onUpgrade` script ever accidentally being run multiple times and executing the same code twice

It might be due to the `<= 20` clause? Can you check if the others just use `<`? I think we need to set this to `<= 19` instead (since the `newVersion` is now `20`)

Comment (HeinousGames) [reply app/src/test/java/co/smartreceipts/android/persistence/database/tables/DistanceTableTest.java]: Yup, I made the update logic incorrect. Thanks for the clarification, I think it's all fixed now.

File: app/src/main/java/co/smartreceipts/android/persistence/database/tables/ReceiptsTable.java [hunk 10]
```diff
@@ -343,6 +347,17 @@ public synchronized void onUpgrade(@NonNull SQLiteDatabase db, int oldVersion, i
             onUpgradeToAddUUID(db, oldVersion);
         }
 
+        if (oldVersion <= 19) { // Added a timezone column to the receipts table
```
Comment (wbaumann) [inline app/src/main/java/co/smartreceipts/android/persistence/database/tables/ReceiptsTable.java]: Btw - this comment is off. Can you fix in the next PR?

File: app/src/main/java/co/smartreceipts/android/persistence/database/tables/TripsTable.java [hunk 11]
```diff
@@ -195,6 +201,20 @@ public synchronized void onUpgrade(@NonNull SQLiteDatabase db, int oldVersion, i
             onUpgradeToAddUUID(db, oldVersion);
         }
 
+        if (oldVersion <= 19) { // Added a timezone column to the receipts table
```
Comment (wbaumann) [inline app/src/main/java/co/smartreceipts/android/persistence/database/tables/TripsTable.java]: Btw - this comment is off. Can you fix in the next PR?

File: app/src/test/java/co/smartreceipts/android/persistence/database/tables/DistanceTableTest.java (hunk 9 above)
Comment (wbaumann) [reply app/src/test/java/co/smartreceipts/android/persistence/database/tables/DistanceTableTest.java]: No worries. That's why we've got the tests :)
